### PR TITLE
changefeedccl: introduce quota to parallelio

### LINKED ISSUE #115536: changefeedccl: test pubsub performance
In https://github.com/cockroachdb/cockroach/issues/111829, we strongly believe that there are performance issues with the pubsub sink. We are likely emitting small batches which is bottlenecking throughput. We must complete these issues to add observability 

- [ ] https://github.com/cockroachdb/cockroach/issues/114027
- [x] https://github.com/cockroachdb/cockroach/issues/114140
- [x] https://github.com/cockroachdb/cockroach/issues/114141

Then, we need to run roachtests and observe the performance of the sink, making the necessary adjustments as needed.

Jira issue: CRDB-34088

### PR DESCRIPTION
**Pull request by jayshrivastava**
Problem:

In this (https://github.com/cockroachdb/cockroach/issues/111829) investigation, it was observed that there was a lot of wasted CPU
performing unions in intsets. This is caused by the `parallelIO` struct which relies
on performing unions to check for conflicting keys. The problem is not with intsets being
slow, but it has to do with how they are used:

`parallelIO` uses a goroutine to both process incoming requests and emit outgoing results.
It accepts incoming requests unconditionally, enqueing them if they cannot be emitted
due to conflicting in flight requests. As outgoing results are processed, each outgoing result
is cross checked with, in the worst case, all enqueued requests to see if they can be emitted.
The cross checking requires unions.

```
incoming request -> request queue -> request handler -> result
                        ^                                 ^
                        | cross check all entries to see  |
                        | if a new request can be emitted |
```

A problem arises when the incoming request queue grows to some critical length where
it significantly slows down the cross checking. This slows down result processing and
ultimately slows down consumption from the request queue. This creates a negative feedback loop
which causes the request queue to grow so large that results take very long to process. This
creates a bottle neck, which throttles the entire changefeed.

See comments https://github.com/cockroachdb/cockroach/issues/115536 for more details.

The request queue is unbounded. The only reason it doesn't cause an OOM is because the incoming
requests are bounded (by the per-changefeed memory limit).

Solution:

This change solves this problem by setting a quota for the maximum events being processed by the
library at the same time. This change sets a size of 128 requests by default. This setting can be
changed using a new cluster setting `changefeed.parallel_io.request_quota`.

Before this change, the API for the parallelio library was very bad. It required the caller to
select on both the request channel and result channel to prevent deadlock. There were also no
public methods. This made it unclear how to properly use the API. This change makes an explicit
API with public methods. However, it keeps the same 2-channel scheme because removing that would
require a larger refactor. This is left as a TODO.

Closes: https://github.com/cockroachdb/cockroach/issues/115536
Release note: None
Epic: None

### CODE REVIEW

**Review by miretskiy — CHANGES_REQUESTED** [review]
> Reviewed 1 of 5 files at r1.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/117692)* status: <span>:shipit:</span> complete! 0 of 0 LGTMs obtained (waiting on @jayshrivastava)

___
*[`pkg/ccl/changefeedccl/parallel_io.go` line 145 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/117692#-NnyrdYxEue1J3QFZ_hV:-NnyrdYxEue1J3QFZ_hW:b9x4mhz) ([raw file](https://github.com/cockroachdb/cockroach/blob/008129265c055e75bdca741cbfc0df39903e4565/pkg/ccl/changefeedccl/parallel_io.go#L145)):*
> ```Go
> }
> 
> func newRequestQueue(sz int) *requestQueue {
> ```

I'm not sure this queue is needed...  Can the caller (batching sink) -- just try to acquire
intpool resource?  Let's discuss.
___
*[`pkg/ccl/changefeedccl/parallel_io.go` line 368 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/117692#-Nny0-mEEsErXGvEn40R:-Nny0-mEEsErXGvEn40S:b-kb7d4h) ([raw file](https://github.com/cockroachdb/cockroach/blob/008129265c055e75bdca741cbfc0df39903e4565/pkg/ccl/changefeedccl/parallel_io.go#L368)):*
> ```Go
> 		}
> 
> 		// Check for incoming requests and ush back on the producer if we have too many pending requests.
> ```

s/ush/push/
___
*[`pkg/ccl/changefeedccl/parallel_io.go` line 370 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/117692#-Nny099aCBOxemLoL3C5:-Nny099aCBOxemLoL3C6:b-f0couu) ([raw file](https://github.com/cockroachdb/cockroach/blob/008129265c055e75bdca741cbfc0df39903e4565/pkg/ccl/changefeedccl/parallel_io.go#L370)):*
> ```Go
> 		// Check for incoming requests and ush back on the producer if we have too many pending requests.
> 		waitForRequest := make(chan struct{})
> 		close(waitForRequest)
> ```

declare (file scope):
```
var closedCh = func() chan struct{} {
  c := make(chan struct{})
  close(c)
  return c
}()
```
and do `waitForRequest = closedCh` to avoid allocating channel on every request.
___
*[`pkg/ccl/changefeedccl/sink.go` line 179 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/117692#-Nny-hXIBtnFVUSX9mme:-Nny-hXIBtnFVUSX9mmf:bdf8peg) ([raw file](https://github.com/cockroachdb/cockroach/blob/12ce879a77d139a51632c0b3b7799137d67e6524/pkg/ccl/changefeedccl/sink.go#L179)):*
> ```Go
> 		" that allows for a higher throughput",
> 	// TODO: delete the original pubsub sink code
> 	util.ConstantWithMetamorphicTestBool("changefeed.new_pubsub_sink.enabled", true),
> ```

Would prefer to have this done in a separate PR.
Also, might make sense to address the TODO.

**Comment by jayshrivastava** [comment]
Still working on this.

**Review by miretskiy — COMMENTED** [review]
> Reviewed 1 of 3 files at r5.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/117692)* status: <span>:shipit:</span> complete! 0 of 0 LGTMs obtained (waiting on @jayshrivastava)

___
*[`pkg/ccl/changefeedccl/batching_sink.go` line 380 at r5](https://reviewable.io/reviews/cockroachdb/cockroach/117692#-Nox59RqEbpq-7LUSm3A:-Nox59Rr7njIFDHFuanM:btkrdw8) ([raw file](https://github.com/cockroachdb/cockroach/blob/1a86fa19356a1dbd89a2d58966576b86354db1a4/pkg/ccl/changefeedccl/batching_sink.go#L380)):*
> ```Go
> 		}
> 
> 		req, send, err := ioEmitter.AdmitRequest(ctx, batchBuffer)
> ```

If I understand correctly, one of the complexities in structuring this code
is the fact that we have to check for "s.doneCh" whenever we try to submit new request;
or wait for them to finish.

Have you considered passing in s.doneCh to IOEmitter so that it can wait for 
ctx.Done and doneCh, and whatever other internal channels it needs to?

Would that result in a more encapsulated code where AdmintRequest would not have
to return a channel, and instead will handle things internall (much like you have here)?
___
*[`pkg/ccl/changefeedccl/batching_sink.go` line 405 at r5](https://reviewable.io/reviews/cockroachdb/cockroach/117692#-NowpCWt5pQYQXyjE7hY:-NowpCWu3KouhkMejbg-:bomgzjb) ([raw file](https://github.com/cockroachdb/cockroach/blob/1a86fa19356a1dbd89a2d58966576b86354db1a4/pkg/ccl/changefeedccl/batching_sink.go#L405)):*
> ```Go
> 		}
> 
> 		// Terminal case. The request was admitted, it must be sent. There
> ```

I view terminal in fatal terms :) 
Maybe rephrase the comment?
___
*[`pkg/ccl/changefeedccl/parallel_io.go` line 43 at r5](https://reviewable.io/reviews/cockroachdb/cockroach/117692#-Nown_ud6Z5njnz5qwCB:-Nown_ud6Z5njnz5qwCC:b-apxbp1) ([raw file](https://github.com/cockroachdb/cockroach/blob/1a86fa19356a1dbd89a2d58966576b86354db1a4/pkg/ccl/changefeedccl/parallel_io.go#L43)):*
> ```Go
> 	ioHandler IOHandler
> 
> 	q         *quotapool.IntPool
> ```

nit: perhaps call it `quota`?

**Review by jayshrivastava — COMMENTED** [review]
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/117692)* status: <span>:shipit:</span> complete! 0 of 0 LGTMs obtained (waiting on @miretskiy)

___
*[`pkg/ccl/changefeedccl/batching_sink.go` line 380 at r5](https://reviewable.io/reviews/cockroachdb/cockroach/117692#-Nox59RqEbpq-7LUSm3A:-Np0vqIM7MswbKL5ys-M:bo342hr) ([raw file](https://github.com/cockroachdb/cockroach/blob/1a86fa19356a1dbd89a2d58966576b86354db1a4/pkg/ccl/changefeedccl/batching_sink.go#L380)):*
<details><summary><i>Previously, miretskiy (Yevgeniy Miretskiy) wrote…</i></summary><blockquote>

If I understand correctly, one of the complexities in structuring this code
is the fact that we have to check for "s.doneCh" whenever we try to submit new request;
or wait for them to finish.

Have you considered passing in s.doneCh to IOEmitter so that it can wait for 
ctx.Done and doneCh, and whatever other internal channels it needs to?

Would that result in a more encapsulated code where AdmintRequest would not have
to return a channel, and instead will handle things internall (much like you have here)?
</blockquote></details>

I had this at one point, but I wasn't a fan of it. I think the better change would be to eliminate the done channel and use context cancellation instead.
___
*[`pkg/ccl/changefeedccl/batching_sink.go` line 405 at r5](https://reviewable.io/reviews/cockroachdb/cockroach/117692#-NowpCWt5pQYQXyjE7hY:-Np0vpQbEvuJL7byRwJZ:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/1a86fa19356a1dbd89a2d58966576b86354db1a4/pkg/ccl/changefeedccl/batching_sink.go#L405)):*
<details><summary><i>Previously, miretskiy (Yevgeniy Miretskiy) wrote…</i></summary><blockquote>

I view terminal in fatal terms :) 
Maybe rephrase the comment?
</blockquote></details>

Done.

**Review by miretskiy — APPROVED** [review]
Giving :lgtm: ; considering adding @rharding6373 to review this code.

> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/117692#-:-Np1J4ecDSa6nH0yfDms:bx872kg)* status: <span>:shipit:</span> complete! 1 of 0 LGTMs obtained (waiting on @jayshrivastava)

**Review by rharding6373 — COMMENTED** [review]
I left a handful of educational questions. Thanks!

> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/117692#-:-Np1ifF-CrFU2BHyJw0O:b4shw3l)* status: <span>:shipit:</span> complete! 1 of 0 LGTMs obtained (waiting on @jayshrivastava and @miretskiy)

___
*[`pkg/ccl/changefeedccl/parallel_io.go` line 129 at r7](https://reviewable.io/reviews/cockroachdb/cockroach/117692#-Np1hz8t4snS1P9V02Co:-Np1hz8t4snS1P9V02Cp:biuu1xt) ([raw file](https://github.com/cockroachdb/cockroach/blob/5612fd58065b3050a28fa8a1fc297f44fb08888f/pkg/ccl/changefeedccl/parallel_io.go#L129)):*
> ```Go
> 	a *quotapool.IntAlloc
> 
> 	// If the request conflicts with in-flight requests, this is the time at
> ```

Clarification question: The time is always updated when placed in the queue, right? Is the part of this comment about conflicts the intended use case for `pendingQueueAdmitTime`?
___
*[`pkg/ccl/changefeedccl/parallel_io.go` line 159 at r7](https://reviewable.io/reviews/cockroachdb/cockroach/117692#-Np1fgfhBo2gt7x6BC3X:-Np1fgfhBo2gt7x6BC3Y:b-fic823) ([raw file](https://github.com/cockroachdb/cockroach/blob/5612fd58065b3050a28fa8a1fc297f44fb08888f/pkg/ccl/changefeedccl/parallel_io.go#L159)):*
> ```Go
> // requestQuota is the number of requests which can be admitted into the
> // parallelio system before blocking the producer.
> var requestQuota = settings.RegisterIntSetting(
> ```

Why does this setting live here instead of changefeedbase/settings.go?
___
*[`pkg/ccl/changefeedccl/parallel_io.go` line 345 at r7](https://reviewable.io/reviews/cockroachdb/cockroach/117692#-Np1cpyjCtCvLpfr9B0V:-Np1cpyjCtCvLpfr9B0W:b-rma6uy) ([raw file](https://github.com/cockroachdb/cockroach/blob/5612fd58065b3050a28fa8a1fc297f44fb08888f/pkg/ccl/changefeedccl/parallel_io.go#L345)):*
> ```Go
> 		// Copy the arrival time for the metrics recorder below.
> 		// Otherwise, it would be possible for res to be admitted to the
> 		// resultCh and freed before we read rec.arrivalTime.
> ```

Off topic but for my education, what is `rec.arrivalTime`?

**Review by jayshrivastava — COMMENTED** [review]
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/117692)* status: <span>:shipit:</span> complete! 1 of 0 LGTMs obtained (waiting on @miretskiy and @rharding6373)

___
*[`pkg/ccl/changefeedccl/parallel_io.go` line 129 at r7](https://reviewable.io/reviews/cockroachdb/cockroach/117692#-Np1hz8t4snS1P9V02Co:-NpM7Rth1RXFrP3-AfSg:b9pn58g) ([raw file](https://github.com/cockroachdb/cockroach/blob/5612fd58065b3050a28fa8a1fc297f44fb08888f/pkg/ccl/changefeedccl/parallel_io.go#L129)):*
<details><summary><i>Previously, rharding6373 (Rachael Harding) wrote…</i></summary><blockquote>

Clarification question: The time is always updated when placed in the queue, right? Is the part of this comment about conflicts the intended use case for `pendingQueueAdmitTime`?
</blockquote></details>

Yes, the time is written when the request is placed in the queue. The pending queue is for requests which cannot immediately be emitted.
___
*[`pkg/ccl/changefeedccl/parallel_io.go` line 159 at r7](https://reviewable.io/reviews/cockroachdb/cockroach/117692#-Np1fgfhBo2gt7x6BC3X:-NpM76cm6vdj1AroVrjA:brhszno) ([raw file](https://github.com/cockroachdb/cockroach/blob/5612fd58065b3050a28fa8a1fc297f44fb08888f/pkg/ccl/changefeedccl/parallel_io.go#L159)):*
<details><summary><i>Previously, rharding6373 (Rachael Harding) wrote…</i></summary><blockquote>

Why does this setting live here instead of changefeedbase/settings.go?
</blockquote></details>

Repeating Yev's answer - we don't have hard rule on this. Public settings can go in settings.go, private ones can go close to their call site / use site.
___
*[`pkg/ccl/changefeedccl/parallel_io.go` line 345 at r7](https://reviewable.io/reviews/cockroachdb/cockroach/117692#-Np1cpyjCtCvLpfr9B0V:-NpM7bC9Dwc-5CWEiDr7:bpuzf6f) ([raw file](https://github.com/cockroachdb/cockroach/blob/5612fd58065b3050a28fa8a1fc297f44fb08888f/pkg/ccl/changefeedccl/parallel_io.go#L345)):*
<details><summary><i>Previously, rharding6373 (Rachael Harding) wrote…</i></summary><blockquote>

Off topic but for my education, what is `rec.arrivalTime`?
</blockquote></details>

Ah. This is a typo. It should be `pendingQueueAdmitTime`. I renamed it in this change. Thanks for catching it.

**Review by jayshrivastava — COMMENTED** [review]
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/117692)* status: <span>:shipit:</span> complete! 1 of 0 LGTMs obtained (waiting on @miretskiy and @rharding6373)

___
*[`pkg/ccl/changefeedccl/parallel_io.go` line 345 at r7](https://reviewable.io/reviews/cockroachdb/cockroach/117692#-Np1cpyjCtCvLpfr9B0V:-NpM8HvX34nvO6GTd58R:bqtwct2) ([raw file](https://github.com/cockroachdb/cockroach/blob/5612fd58065b3050a28fa8a1fc297f44fb08888f/pkg/ccl/changefeedccl/parallel_io.go#L345)):*
<details><summary><i>Previously, jayshrivastava (Jayant) wrote…</i></summary><blockquote>

Ah. This is a typo. It should be `pendingQueueAdmitTime`. I renamed it in this change. Thanks for catching it.
</blockquote></details>

Correction* it should be resultTime

**Comment by jayshrivastava** [comment]
The throughput of the sink looks good 
image

The fact that the keys are capping at 3000 likely means that throttling is kicking in.
image
 It's very spiky... Probably because of the config below. 

image

**Comment by jayshrivastava** [comment]
https://grafana.testeng.crdb.io/d/crdb-console-changefeeds-byname/crdb-console-changefeeds-by-test-run-and-name?from=1706565131061&to=1706566358334&var-DS_PROMETHEUS=Prometheus&var-test_run_id=jayants-1706564753&var-test_name=cdc-pubsub-sink&var-cluster=&orgId=1
